### ISSUE TITLE
no std mode

### ISSUE THREAD

**Issue by artit91** [wasmerio/wasmer#217]
I want to run wasmer on bare metal. Is it on the roadmap?

**Comment by lachlansneff** [comment]
Hi!

Running wasmer on bare-metal is on the long-term roadmap, but it's not a current priority. The main blockers are that wasmer requires the ability to map memory with various protections and that it needs to be able to hook into os-level signals/exceptions.

I have a special place in my heart for rust running without an operating system, so I'd love to see this happen, however.

If you'd like to work on this feature, we'd be happy to accept any contributions and would provide help and advice! Contact us on the spectrum channel if you'd like to pursue this!

**Comment by artit91** [comment]
Hi,

I think that's why WebAssembly is sandboxed by default and you can bind your context from outside.

The current (unix) implementation depends on various libc functions that shall be just another optional context to bind. If you want to write a 3D tech demo using typescript for example that is using only a videocore chip and nothing else, you can't use this awesome lib sadly.

**Comment by syrusakbary** [comment]
Agreed! We would love to have no-std support, as it will be useful for a lot of different scenarios like the one you commented. Right now we are prioritizing other cases like improving speed of the runtime or help the usage of WebAssembly runtime from other languages.

However, we would love to support you on the no-std quest! Is there anything we can do for helping you to achieve that?

**Comment by assafmo** [comment]
Hi, any progress on `no_std`? :pray:

**Comment by assafmo** [comment]
We @enigmampc need `no_std` in order to run WASM inside an SGX enclave.
Currently, our only solution is to use https://github.com/paritytech/wasmi which is slower, has no built-in gas/metering support and has a slightly worse UX IMO.
So `no_std` would come in handy.

**Comment by Hywan** [comment]
What backend would you like to use in `no_std` mode? It could help us to set priority.

**Comment by assafmo** [comment]
We are using [CosmWasm](https://github.com/CosmWasm/cosmwasm), and I believe they are using `wasmer-singlepass-backend`.
But because this doesn't support `no_std` we are working on replacing `wasmer` with `wasmi` inside of `CosmWasm`.
If wasmer would support `no_std` we'd use it as it'll save us a lot of work.

**Comment by syrusakbary** [comment]
@assafmo we would love to support ‘no_std’ in Wasmer.

Right now we are focusing our efforts in making the codebase simpler, but we can guide and help with any contributions.
Would you like to contribute in making Wasmer compatible with ‘no_std’?

**Comment by assafmo** [comment]
Cc @ethanfrey @webmaster128

@syrusakbary We would love to but let me check if we have the capacity right now. :blush:

**Comment by stevemk14ebr** [comment]
Any updates? This would be really fantastic for embedded platforms and kernels. Wasmer seems one of the most complete wasm engines.

**Comment by Hywan** [comment]
I would say that maybe 50% of our crates can be used in `no_std`. I don't know how much work would be required to me it up to 100%. Anyone would like to help on that front? I can coordinate the project.

**Comment by 1c3t3a** [comment]
- I would like to help - if I can. I'm currently preparing for my bachelor thesis about virtualization with WASM in the embedded area, where I would kinda depend on a `no_std` wasmer version.

**Comment by syrusakbary** [comment]
Awesome.

I think there was some work done on a PR that you can probably leverage on:
https://github.com/wasmerio/wasmer/pull/2168

Feel free to create a side PR with the progress so nostd support can be merged upstream :)

**Comment by 1c3t3a** [comment]
Great thanks for the quick response! I am going to head over to the PR and have al look!

**Comment by ethindp** [comment]
Any update on this? For mmap/mprotect you could just have a trait for the host to define its own functions for that functionality. Then you wouldn't need to care how the host creates executable pages in memory. Same for signals -- just use a trait for handler functions that either raises the signal or whatever.

What else is blocking this? Other than mmap/mprotect and signals?

**Comment by nerdyguy87** [comment]
> 

Hello. For our project we are looking to use Wasmr on bare-metal for aarch64 and RiscV64. We would need to be able to use the cranelift backend  in no_std mode. Our team is disheartened that no_std is set as a low priority and have reluctantly been looking into alternatives like wasmi. Could we please request higher priority support with this?

**Comment by ethindp** [comment]
I'm looking to use the cranelift backend as well. Its incredibly disheartening to see that all the *actually good* Wasm implementations put no_std support as a very low priority feature. Wasmi is the only no_std implementation available, and its fallen extremely far behind and is slow (and I also find the API convoluted). (It doesn't help that my research shows that nobody seems to publicly use it either, which means that, for all intents and purposes, people who use it are left to their own devices.)

**Comment by 1c3t3a** [comment]
I agree, but to be fair, reaching no_std support is a fairly big step and a lot of work. Also maintaining it can be hard, as every dependency potentially breaks no_std. @nerdyguy87 did you take a look at Wasm-Micro-runtime? It has JiT, AoT and Interpreter support and runs on several systems (Zephyr, vxworks, ESP-IDF, SGX). 

Generally I think support for Espressif's [Rust standard library implementation](https://github.com/ivmarkov/rust-esp32-std-demo) is easier to reach, as it mostly requires changes to libc and the memory mapping code. Then it's at least possible to run Wasmer on the Espressif chips.

**Comment by stevemk14ebr** [comment]
Strong disagree with the espressif idea. General no_std support is much better, in my use case I need no_std to run wasm in the windows kernel. I do not care about targeting specific chip and that would only solve a tiny sliver of all use cases while still being a maintenance burden

**Comment by 1c3t3a** [comment]
As far as I can tell the general idea is to compile nearly any rust std dependendant code. That wouldn't introduce a maintenance burden at wasmers.

**Comment by stevemk14ebr** [comment]
bad bot

**Comment by artit91** [comment]
Some point next year I will have a few months to look into this properly, but I doubt it can be done as a single PR. I currently use a bare minimal linux to run wasmer that kind of defeats the point.

**Comment by Lucky4Luuk** [comment]
Any more updates on this? I'd to love to use wasmer in a no_std environment, with alloc support, regardless of the backend. The performance gains over wasmi would be very useful

**Comment by assafmo** [comment]
Hey guys, any update on this?

**Comment by stevemk14ebr** [comment]
Disable this trash bot already

**Comment by stevemk14ebr** [comment]
Has the status of this changed at all with the release of 5.0? https://wasmer.io/posts/introducing-wasmer-v5 states:

"Wasmi: one of the most optimal Rust WebAssembly interpreters. Ideal if you want to use Wasm in nostd environments or blockchains with a pure Rust codebase (wasmi feature)"

Wasmi is the runtime I switched to instead of wasmer because it supports no_std completely. Now that wasmer has the wasmi backend, has this issue progressed? I do not yet see a no_std feature flag in the wasmer cargo.toml
